src/System.ComponentModel.Annotations/tests/MaxLengthAttributeTests.cs

```diff
@@ -30,9 +31,14 @@ protected static IEnumerable<object[]> ValidValues_ICollection()
             yield return new object[] { new MaxLengthAttribute(-1), new Collection<int>(new int[20]) };
             yield return new object[] { new MaxLengthAttribute(15), new Collection<string>(new string[14]) };
             yield return new object[] { new MaxLengthAttribute(16), new Collection<string>(new string[16]) };
+
             yield return new object[] { new MaxLengthAttribute(-1), new List<int>(new int[20]) };
             yield return new object[] { new MaxLengthAttribute(15), new List<string>(new string[14]) };
             yield return new object[] { new MaxLengthAttribute(16), new List<string>(new string[16]) };
+
+            yield return new object[] { new MaxLengthAttribute(-1), new HashSet<int>(Enumerable.Range(1, 20)) };
+            yield return new object[] { new MaxLengthAttribute(15), new HashSet<string>(Enumerable.Range(1, 14).Select(i => i.ToString())) };
+            yield return new object[] { new MaxLengthAttribute(16), new HashSet<string>(Enumerable.Range(1, 16).Select(i => i.ToString())) };
```

Comment:
Make sure there are tests for:
* Types that are `ICollection` but not `ICollection<T>` and vice versa
* Types that implement `ICollection<T>` multiple times
* If we do pattern matching, then types that have a Count but are not ICollection.

Reply:
> Types that are ICollection but not ICollection<T> **(a)** and vice versa **(b)**

a. Since we rely on the `Count` property this is redundant. I could add `ArrayList` which will fall through `ICollection`.
b. That's the purpose of `HashSet<T>` I've added to the tests

> Types that implement ICollection<T> multiple times

Again since `ICollection<T>` is out of the picture this is redundant.

> If we do pattern matching, then types that have a Count but are not ICollection.

Again, `HashSet<T>` satisfies this test case.

See my latest commit: https://github.com/dotnet/corefx/pull/23664/commits/6353a8f4d634b4d82d24626621c7f0cae1e93758

Reply:
@weitzhandler I recognize that some of these will hit the same code path multiple times, but they are still useful in a couple of ways:
* They make sure that our reasoning about which cases are covered is correct
* They act as regression tests so that if the implementation changes in the future (for example, if ICountable or something like it happens) then the existing behaviors are still tested.

Reply:
Done https://github.com/dotnet/corefx/pull/23664/commits/4cba646a9db0bb91e6aa88bbf0594b71d33844f5.

Reply:
@ajcvickers 
Done https://github.com/dotnet/corefx/pull/23664/commits/4cba646a9db0bb91e6aa88bbf0594b71d33844f5.
The multi `ICollection<T>` looks totally redundant to me but I implemented it anyway per your request.